src/main/java/com/openshift/internal/restclient/ApiTypeMapper.java

```diff
@@ -0,0 +1,389 @@
+import java.util.stream.Stream;
+
+import org.apache.commons.lang.StringUtils;
+import org.jboss.dmr.ModelNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.openshift.internal.restclient.http.HttpClientException;
+import com.openshift.internal.restclient.http.NotFoundException;
+import com.openshift.internal.restclient.model.properties.ResourcePropertyKeys;
+import com.openshift.internal.util.JBossDmrExtentions;
+import com.openshift.restclient.IApiTypeMapper;
+import com.openshift.restclient.OpenShiftException;
+import com.openshift.restclient.ResourceKind;
+import com.openshift.restclient.http.IHttpClient;
+import com.openshift.restclient.http.IHttpConstants;
+import com.openshift.restclient.model.IResource;
+
+/**
+ * Typemapper to determine the endpoints for
+ * various openshift resources
+ * @author jeff.cantrill
+ *
+ */
+public class ApiTypeMapper implements IApiTypeMapper, ResourcePropertyKeys{
+	
+	private static final String FWD_SLASH = "/";
+	private static final String KUBE_API = "api";
+	private static final String OS_API = "oapi";
+	private static final String API_GROUPS_API = "apis";
+	private static final Logger LOGGER = LoggerFactory.getLogger(ApiTypeMapper.class);
+	private final String baseUrl;
+	private final IHttpClient client;
+	private List<VersionedApiResource> resourceEndpoints;
+	private Map<String, String> preferedVersion = new HashMap<>(2);
+	
+	public ApiTypeMapper(String baseUrl, IHttpClient client) {
+		this.baseUrl = baseUrl;
+		this.client = client;
+		preferedVersion.put(KUBE_API, KubernetesAPIVersion.v1.toString());
+		preferedVersion.put(OS_API, OpenShiftAPIVersion.v1.toString());
+	}
+
+	@Override
+	public boolean isSupported(IResource resource) {
+		return isSupported(resource.getApiVersion(), resource.getKind());
+	}
+	
+	@Override
+	public IVersionedApiResource getEndpointFor(String apiVersion, String kind) {
+		IVersionedApiResource apiresource = endpointFor(apiVersion, kind);
+		if(apiresource == null) {
+			throw new OpenShiftException("No endpoint found for kind %s", kind);
+		}
+		return apiresource;
+	}
+	
+	private IVersionedApiResource endpointFor(String version, String kind) {
+		String[] split = StringUtils.isBlank(version) ? new String [] {} : version.split(FWD_SLASH);
+		Optional<IVersionedApiResource> result = null;
+		if(split.length <=1) {
+			result = Stream.of(KUBE_API, OS_API)
+					.map(api->formatEndpointFor(api, (split.length == 0 ? preferedVersion.get(api) : split[0]), kind))
+					.filter(e->resourceEndpoints.contains(e))
+					.findFirst();
+		}else{
+			result = Optional.of(formatEndpointFor(API_GROUPS_API, version, ResourceKind.pluralize(kind, true, true)));
+		}
+		if(result.isPresent()) {
+			int index = resourceEndpoints.indexOf(result.get());
+			if(index > -1) {
+				return resourceEndpoints.get(index);
+			}
+		}
+		return null;
+	}
+
+
+	@Override
+	public boolean isSupported(String version, String kind) {
+		init();
```

Comment:
apparently #isSupported is required to be called in the 1st place (since it calls #init). Shouldnt we therefore assert in the other methods that #init was called?